syntax = "proto3";

package carcrash;

import "google/protobuf/empty.proto";

message Point {
    int32 x = 1;
    int32 y = 2;
}

message Id {
    uint32 id = 1;
}

message Rectangle {
    enum Color {
        Black = 0;
        White = 1;
        Red = 2;
        Green = 3;
        Blue = 4;
        Yellow = 5;
        Magenta = 6;
        Cyan = 7;
        Transparent = 8;
    }

    uint32 id = 1;
    Point position = 2;
    float angle = 3;
    Point dimension = 4;
    Color color = 5;
}

message Vehicles {
    repeated Rectangle rectangles = 1;
}

message VehicleWithId {
    Id id = 1;
    Rectangle rectangle = 2;
}

service CarCrash {
    rpc GetNewId (google.protobuf.Empty) returns (Id) {
    }
    rpc GetVehicles (google.protobuf.Empty) returns (Vehicles) {
    }
    rpc UpdateVehicle (VehicleWithId) returns (google.protobuf.Empty) {
    }
    rpc RegisterVehicle (VehicleWithId) returns (Id) {
    }
    rpc UnregisterVehicle (Id) returns (Id) { // get id to Unregister, returns last Id
    }
}

